theano/sandbox/cuda/corr_gemm.cu

```diff
@@ -84,75 +85,90 @@ __global__ void im2col_kernel(const int n, const float* data_im,
   int num_kernels = channels * height_col * width_col;
   im2col_kernel<<<GET_BLOCKS(num_kernels),
                   CUDA_NUM_THREADS>>>(
-      num_kernels, data_im, height, width, kernel_h, kernel_w, pad_h,
-      pad_w, stride_h, stride_w, height_col,
+      num_kernels, data_im, height, width, kernel_h, kernel_w,
+      dilation_h, dilation_w, pad_h, pad_w, stride_h, stride_w, height_col,
       width_col, data_col);
 }
 
 __global__ void col2im_kernel(const int n, const float* data_col,
     const int height, const int width, const int channels,
     const int patch_h, const int patch_w,
+    const int dilation_h, const int dilation_w,
     const int pad_h, const int pad_w,
     const int stride_h, const int stride_w,
     const int height_col, const int width_col,
     float* data_im) {
+  int dil_patch_h = (patch_h-1)*dilation_h + 1;
+  int dil_patch_w = (patch_w-1)*dilation_w + 1;
   CUDA_KERNEL_LOOP(index, n) {
     float val = 0;
     int w = index % width + pad_w;
     int h = (index / width) % height + pad_h;
     int c = index / (width * height);
     // compute the start and end of the output
-    int w_col_start = (w < patch_w) ? 0 : (w - patch_w) / stride_w + 1;
+    int w_col_start = (w < dil_patch_w) ? 0 : (w - dil_patch_w) / stride_w + 1;
     int w_col_end = min(w / stride_w + 1, width_col);
-    int h_col_start = (h < patch_h) ? 0 : (h - patch_h) / stride_h + 1;
+    int h_col_start = (h < dil_patch_h) ? 0 : (h - dil_patch_h) / stride_h + 1;
     int h_col_end = min(h / stride_h + 1, height_col);
-    /*
+
     for (int h_col = h_col_start; h_col < h_col_end; ++h_col) {
       for (int w_col = w_col_start; w_col < w_col_end; ++w_col) {
         // the col location: [c * width * height + h_out, w_out]
-        int c_col = c * patch_h * patch_w + (h - h_col * stride_h) * ksize
-            + (w - w_col * stride_w);
-        val += data_col[(c_col * height_col + h_col) * width_col + w_col];
+	int h_k = h - h_col * stride_h;
+	int w_k = w - w_col * stride_w;
+	if (h_k % dilation_h == 0 && w_k % dilation_w == 0) {
+          h_k /= dilation_h;
+          w_k /= dilation_w;
+	  int c_col = c * patch_h * patch_w + h_k * patch_w
+            + w_k;
+	  val += data_col[(c_col * height_col + h_col) * width_col + w_col];
+	 }
       }
     }
-    */
+    
     // equivalent implementation
-    int offset =
-        (c * patch_h * patch_w + h * patch_w + w) * height_col * width_col;
+    /*int offset =
+      (c * patch_h * patch_w + h * patch_w + w) * height_col * width_col;
     int coeff_h_col = (1 - stride_h * patch_w * height_col) * width_col;
     int coeff_w_col = (1 - stride_w * height_col * width_col);
     for (int h_col = h_col_start; h_col < h_col_end; ++h_col) {
       for (int w_col = w_col_start; w_col < w_col_end; ++w_col) {
-        val += data_col[offset + h_col * coeff_h_col + w_col * coeff_w_col];
+	val += data_col[offset + h_col * coeff_h_col + w_col * coeff_w_col];
       }
-    }
+      }*/
```

Comment:
Why these indentation changes? We may want to drop the "equivalent implementation" now that it's not equivalent any more.